### PR TITLE
Feature/get salaries in cities of country

### PR DESCRIPTION
**Pull request by HaidyAbuGom3a**

### CODE REVIEW

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    private val dataSource: CostOfLivingDataSource
+    ) {
+
+    fun execute(countryName:String):List<Pair<String,Float?>> {
+        return if (checkIfCountryIsValid(countryName))
+            dataSource.getAllCitiesData()
+                .filter{excludeNullSalariesAndLowQualityData(it) && it.country.lowercase() == countryName.lowercase()}
+                .map{Pair(it.cityName,it.averageMonthlyNetSalaryAfterTax)}
+        else
+            throw IOException("Invalid country name")
```
Comment (ma28alhams) [inline src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:16]:
```suggestion
            throw Exception("Invalid country name")
```
IOExceptions just for exceptions when accessing the **files and directories**,

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package interactor
+
+import model.CityEntity
+import java.io.IOException
+
+class GetSalaryAverageAndCitiesNamesInCountryInteractor(
+    private val dataSource: CostOfLivingDataSource
+    ) {
+
+    fun execute(countryName:String):List<Pair<String,Float?>> {
```
Comment (ma28alhams) [inline src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:10]: You have execluded the null salaries, so you have not to make it return **Nalluable** 
```suggestion
    fun execute(countryName:String):List<Pair<String,Float>> {
```
Make it :
`List<Pair<String,**Float**>>`

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        return if (checkIfCountryIsValid(countryName))
+            dataSource.getAllCitiesData()
+                .filter{excludeNullSalariesAndLowQualityData(it) && it.country.lowercase() == countryName.lowercase()}
```
Comment (ma28alhams) [inline src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:13]: Execlude Low quality data and null salaries first, then check if the country name is valid, and you should make the **checkIfCountryIsValid** in the filter method
```suggestion
          dataSource.getAllCitiesData()
          .filter{excludeNullSalariesAndLowQualityData(it) && checkIfCountryIsValid(it.countryName)}
```

**Review by ma28alhams — CHANGES_REQUESTED** [review]

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt (hunk 1 above)
Comment (HaidyAbuGom3a) [reply src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt]: done

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt (hunk 2 above)
Comment (HaidyAbuGom3a) [reply src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt]: I did that before but i was getting error. I solved it with !! in the second value of pair in the return.
done!

File: src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt (hunk 3 above)
Comment (HaidyAbuGom3a) [reply src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:13]: I tried this but it gives me wrong tests

Comment (ma28alhams) [reply src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:13]: No, it's okay, I'm sorry, your code is more efficient

Comment (ma28alhams) [reply src/main/kotlin/interactor/GetSalaryAverageAndCitiesNamesInCountryInteractor.kt:13]: First, it checks if the country name is valid, then and ONLY then should continue.

**Pull request merged**